File: domain-persistence-modeling/java-spring-boot/src/test/java/x/scratch/parent/PersistedParentTest.java
package x.scratch.parent;

import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;
import x.scratch.DomainChangedEvent;
import x.scratch.DomainException;
import x.scratch.PersistableDomain.UpsertedDomainResult;
import x.scratch.TestListener;
import x.scratch.child.Child;
import x.scratch.child.ChildChangedEvent;
import x.scratch.child.ChildFactory;
import x.scratch.child.ChildResource;

import java.util.List;

import static java.util.Collections.emptySet;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
import static org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace.NONE;
import static x.scratch.parent.MutableParent.Helper.assign;
import static x.scratch.parent.MutableParent.Helper.unassign;

@AutoConfigureTestDatabase(replace = NONE)
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@SpringBootTest
@TestInstance(PER_CLASS)
@Transactional
class PersistedParentTest {
    private static final String parentNaturalId = "a";
    private static final String childNaturalId = "p";

    private final ParentFactory parents;
    private final ChildFactory children;
    private final TestListener<DomainChangedEvent<?>> testListener;

    @Test
    void shouldCreateNew() {
        final var found = parents.findExistingOrCreateNew(parentNaturalId);

        assertThat(found).isEqualTo(parents.createNew(parentNaturalId));
        assertThat(found.getChildren()).isEmpty();
    }

    @Test
    void shouldFindExisting() {
        final var saved = newSavedParent();

        final var found = parents.findExistingOrCreateNew(parentNaturalId);

        assertThat(found).isEqualTo(saved);
        assertThat(found.getChildren()).isEmpty();
    }

    @Test
    void shouldRoundTrip() {
        final var unsaved = parents.createNew(parentNaturalId);

        assertThat(unsaved.getVersion()).isEqualTo(0);
        assertThat(events()).isEmpty();

        final var saved = unsaved.save();

        assertThat(parents.all()).hasSize(1);
        assertThat(unsaved.getVersion()).isEqualTo(1);
        assertThat(saved).isEqualTo(UpsertedDomainResult.of(unsaved, true));
        assertThat(events()).containsExactly(new ParentChangedEvent(
                null,
                new ParentResource(parentNaturalId, null, 1)));

        assertThat(currentPersistedParent()).isEqualTo(unsaved);
    }

    @Test
    void shouldDetectNoChanges() {
        final var original = newSavedParent();
        final var resaved = original.save();

        assertThat(resaved).isEqualTo(
                UpsertedDomainResult.of(original, false));
        assertThat(events()).isEmpty();
    }

    @Test
    void shouldMutate() {
        final var original = newSavedParent();

        assertThat(original.isChanged()).isFalse();

        final var value = "FOOBAR";
        final var modified = original.update(it -> it.setValue(value));

        assertThat(modified).isEqualTo(original);
        assertThat(original.isChanged()).isTrue();
        assertThat(original.getValue()).isEqualTo(value);
        assertThat(events()).isEmpty();

        original.save();

        assertThat(original.isChanged()).isFalse();
        assertThat(events()).containsExactly(new ParentChangedEvent(
                new ParentResource(parentNaturalId, null, 1),
                new ParentResource(parentNaturalId, value, 2)));
    }

    @Test
    void shouldMutateChildren() {
        final var parent = newSavedParent();
        final var child = newSavedChild();

        parent.update(assign(child)).save();
        testListener.reset();

        final var value = "FOOBAR";
        // Silly example :)
        parent.update(it ->
                it.getChildren().forEach(itt ->
                        itt.update(ittt ->
                                ittt.setValue(value))));
        parent.save();

        assertThat(currentPersistedChild().getValue()).isEqualTo(value);

        assertThat(events()).containsExactly(
                new ChildChangedEvent(
                        new ChildResource(childNaturalId, parentNaturalId,
                                null, emptySet(), 2),
                        new ChildResource(childNaturalId, parentNaturalId,
                                value, emptySet(), 3)),
                new ParentChangedEvent(
                        new ParentResource(parentNaturalId, null, 2),
                        new ParentResource(parentNaturalId, null, 3)));
    }

    @Test
    void shouldDelete() {
        final var existing = newSavedParent();

        existing.delete();

        assertThat(parents.all()).isEmpty();
        assertThatThrownBy(existing::getVersion)
                .isInstanceOf(NullPointerException.class);
        assertThat(events()).containsExactly(new ParentChangedEvent(
                new ParentResource(parentNaturalId, null, 1),
                null));
    }

    @Test
    void shouldNotDelete() {
        final var parent = newSavedParent();
        final var child = newSavedChild();

        parent.update(assign(child));

        assertThatThrownBy(parent::delete)
                .isInstanceOf(DomainException.class);
    }

    @Test
    void shouldNotAssignAlreadyAssignedChild() {
        final var parent = newSavedParent();
        final var child = newSavedChild();

        parent.update(assign(child)).save();

        assertThatThrownBy(() -> parent.update(assign(child)).save())
                .isInstanceOf(DomainException.class);
    }

    @Test
    void shouldAssignAndUnassignChild() {
        final var parent = newSavedParent();
        final var child = newSavedChild();

        assertThat(parent.getChildren()).isEmpty();

        final var childAssigned = parent
                .update(assign(child))
                .save().getDomain();

        assertThat(parent.getChildren()).containsExactly(child);
        assertThat(childAssigned.getVersion()).isEqualTo(2);
        assertThat(currentPersistedChild().getParentNaturalId())
                .isEqualTo(parentNaturalId);
        assertThat(events()).containsExactly(
                new ChildChangedEvent(
                        new ChildResource(childNaturalId, null, null,
                                emptySet(), 1),
                        new ChildResource(childNaturalId, parentNaturalId,
                                null, emptySet(), 2)),
                new ParentChangedEvent(
                        new ParentResource(parentNaturalId, null, 1),
                        new ParentResource(parentNaturalId, null, 2)));

        final var childUnassigned = parent
                .update(unassign(child))
                .save().getDomain();

        assertThat(parent.getChildren()).isEmpty();
        assertThat(childUnassigned.getVersion()).isEqualTo(3);
        assertThat(currentPersistedChild().getParentNaturalId()).isNull();
        assertThat(events()).containsExactly(
                new ChildChangedEvent(
                        new ChildResource(childNaturalId, parentNaturalId,
                                null, emptySet(), 2),
                        new ChildResource(childNaturalId, null, null,
                                emptySet(), 3)),
                new ParentChangedEvent(
                        new ParentResource(parentNaturalId, null, 2),
                        new ParentResource(parentNaturalId, null, 3)));
    }

    private Child newSavedChild() {
        final var saved = children.createNewUnassigned(childNaturalId).save();
        assertThat(saved.isChanged()).isTrue();
        final var child = saved.getDomain();
        testListener.reset();
        return child;
    }

    private Child currentPersistedChild() {
        return children.findExisting(childNaturalId).orElseThrow();
    }

    private Parent newSavedParent() {
        final var saved = parents.createNew(parentNaturalId).save();
        assertThat(saved.isChanged()).isTrue();
        final var parent = saved.getDomain();
        testListener.reset();
        return parent;
    }

    private Parent currentPersistedParent() {
        return parents.findExisting(parentNaturalId).orElseThrow();
    }

    private List<DomainChangedEvent<?>> events() {
        return testListener.events();
    }
}

File: spring-boot-domain-persistence-modeling-kotlin/src/test/kotlin/x/domainpersistencemodeling/PersistedParentTest.kt
package x.domainpersistencemodeling

import ch.tutteli.atrium.api.cc.en_GB.containsExactly
import ch.tutteli.atrium.api.cc.en_GB.hasSize
import ch.tutteli.atrium.api.cc.en_GB.isEmpty
import ch.tutteli.atrium.api.cc.en_GB.toBe
import ch.tutteli.atrium.api.cc.en_GB.toThrow
import ch.tutteli.atrium.verbs.expect
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase
import org.springframework.boot.test.autoconfigure.jdbc.AutoConfigureTestDatabase.Replace.NONE
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional
import x.domainpersistencemodeling.PersistableDomain.UpsertedDomainResult

@AutoConfigureTestDatabase(replace = NONE)
@SpringBootTest
@Transactional
internal open class PersistedParentTest @Autowired constructor(
        private val parents: ParentFactory,
        private val children: ChildFactory,
        private val testListener: TestListener<ParentChangedEvent>) {
    companion object {
        private const val parentNaturalId = "a"
        private const val childNaturalId = "p"
    }

    @Test
    fun shouldCreateNew() {
        val found = parents.findExistingOrCreateNew(parentNaturalId)

        expect(found).toBe(parents.createNew(parentNaturalId))
        expect(found.children).isEmpty()
    }

    @Test
    fun shouldFindExisting() {
        val saved = newSavedParent()

        val found = parents.findExistingOrCreateNew(parentNaturalId)

        expect(found).toBe(saved)
        expect(found.children).isEmpty()
    }

    @Test
    fun shouldRoundTrip() {
        val unsaved = parents.createNew(parentNaturalId)

        expect(unsaved.version).toBe(0)
        testListener.expectNext.isEmpty()

        val saved = unsaved.save()

        expect(parents.all().toList()).hasSize(1)
        expect(unsaved.version).toBe(1)
        expect(saved).toBe(UpsertedDomainResult(unsaved, true))
        testListener.expectNext.containsExactly(ParentChangedEvent(
                null,
                ParentResource(parentNaturalId, null, 1)))

        expect(currentPersistedParent()).toBe(unsaved)
    }

    @Test
    fun shouldDetectNoChanges() {
        val original = newSavedParent()
        val resaved = original.save()

        expect(resaved).toBe(UpsertedDomainResult(original, false))
        testListener.expectNext.isEmpty()
    }

    @Test
    fun shouldMutate() {
        val original = newSavedParent()

        expect(original.changed).toBe(false)

        val value = "FOOBAR"
        val modified = original.update {
            this.value = value
        }

        expect(modified).toBe(original)
        expect(original.changed).toBe(true)
        expect(original.value).toBe(value)
        testListener.expectNext.isEmpty()

        original.save()

        expect(original.changed).toBe(false)
        testListener.expectNext.containsExactly(ParentChangedEvent(
                ParentResource(parentNaturalId, null, 1),
                ParentResource(parentNaturalId, value, 2)))
    }

    @Test
    fun shouldMutateChildren() {
        val parent = newSavedParent()
        val child = newSavedChild()

        parent.update {
            assign(child)
        }.save()
        testListener.reset()

        val value = "FOOBAR"
        // Silly example :)
        parent.update {
            children.forEach {
                it.update {
                    this.value = value
                }
            }
        }.save()

        expect(currentPersistedChild().value).toBe(value)

        testListener.expectNext.containsExactly(
                ChildChangedEvent(
                        ChildResource(childNaturalId, parentNaturalId,
                                null, emptySet(), 2),
                        ChildResource(childNaturalId, parentNaturalId,
                                value, emptySet(), 3)),
                ParentChangedEvent(
                        ParentResource(parentNaturalId, null, 2),
                        ParentResource(parentNaturalId, null, 3)))
    }

    @Test
    fun shouldDelete() {
        val existing = newSavedParent()

        existing.delete()

        expect(parents.all().toList()).isEmpty()
        expect {
            existing.version
        }.toThrow<DomainException> { }
        testListener.expectNext.containsExactly(ParentChangedEvent(
                ParentResource(parentNaturalId, null, 1),
                null))
    }

    @Test
    fun shouldNotDelete() {
        val parent = newSavedParent()
        val child = newSavedChild()

        parent.update {
            assign(child)
        }

        expect {
            parent.delete()
        }.toThrow<DomainException> { }
    }

    @Test
    fun shouldNotAssignAlreadyAssignedChild() {
        val parent = newSavedParent()
        val child = newSavedChild()

        parent.update {
            assign(child)
        }.save()

        expect {
            parent.update {
                assign(child)
            }.save()
        }.toThrow<DomainException> { }
    }

    @Test
    fun shouldAssignAndUnassignChild() {
        val parent = newSavedParent()
        val child = newSavedChild()

        expect(parent.children).isEmpty()

        val parentAssignedWithChild = parent.update {
            assign(child)
        }.save().domain

        expect(parent.children).containsExactly(child)
        expect(parentAssignedWithChild.version).toBe(2)
        expect(currentPersistedChild().parentNaturalId)
                .toBe(parentNaturalId)
        testListener.expectNext.containsExactly(
                ChildChangedEvent(
                        ChildResource(childNaturalId, null, null,
                                emptySet(), 1),
                        ChildResource(childNaturalId, parentNaturalId, null,
                                emptySet(), 2)),
                ParentChangedEvent(
                        ParentResource(parentNaturalId, null, 1),
                        ParentResource(parentNaturalId, null, 2)))

        val childUnassigned = parent.update {
            unassign(child)
        }.save().domain

        expect(parent.children).isEmpty()
        expect(childUnassigned.version).toBe(3)
        expect(currentPersistedChild().parentNaturalId).toBe(null)
        testListener.expectNext.containsExactly(
                ChildChangedEvent(
                        ChildResource(childNaturalId, parentNaturalId, null,
                                emptySet(), 2),
                        ChildResource(childNaturalId, null, null,
                                emptySet(), 3)),
                ParentChangedEvent(
                        ParentResource(parentNaturalId, null, 2),
                        ParentResource(parentNaturalId, null, 3)))
    }

    private fun newSavedChild(): Child {
        val saved = children.createNew(childNaturalId).save()
        expect(saved.changed).toBe(true)
        val child = saved.domain
        testListener.reset()
        return child
    }

    private fun currentPersistedChild() =
            children.findExisting(childNaturalId)!!

    private fun newSavedParent(): Parent {
        val saved = parents.createNew(parentNaturalId).save()
        expect(saved.changed).toBe(true)
        val parent = saved.domain
        testListener.reset()
        return parent
    }

    private fun currentPersistedParent() =
            parents.findExisting(parentNaturalId)!!
}
